pkg/controllermanager/controller/managedseedset/add.go

```diff
@@ -0,0 +1,372 @@
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+
+	if r.Actuator == nil {
+		replicaFactory := ReplicaFactoryFunc(NewReplica)
+		replicaGetter := NewReplicaGetter(r.Client, mgr.GetAPIReader(), replicaFactory)
+		r.Actuator = NewActuator(r.Client, replicaGetter, replicaFactory, &r.Config, mgr.GetEventRecorderFor(ControllerName+"-controller"))
+	}
+
+	c, err := builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&seedmanagementv1alpha1.ManagedSeedSet{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: pointer.IntDeref(r.Config.ConcurrentSyncs, 0),
+			RecoverPanic:            true,
+		}).
+		Build(r)
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &gardencorev1beta1.Shoot{}},
+		&handler.EnqueueRequestForOwner{
+			OwnerType:    &seedmanagementv1alpha1.ManagedSeedSet{},
+			IsController: true,
+		},
+		r.ShootPredicate(),
+	); err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &seedmanagementv1alpha1.ManagedSeed{}},
+		&handler.EnqueueRequestForOwner{
+			OwnerType:    &seedmanagementv1alpha1.ManagedSeedSet{},
+			IsController: true,
+		},
+		r.ManagedSeedPredicate(),
+	); err != nil {
+		return err
+	}
+
+	return c.Watch(
+		&source.Kind{Type: &gardencorev1beta1.Seed{}},
+		mapper.EnqueueRequestsFrom(mapper.MapFunc(r.MapSeedToManagedSeedSet), mapper.UpdateWithNew, c.GetLogger()),
+		r.SeedPredicate(),
+	)
+}
+
+// ShootPredicate returns the predicate for Shoot events.
+func (r *Reconciler) ShootPredicate() predicate.Predicate {
+	return &shootPredicate{}
+}
+
+type shootPredicate struct {
+	ctx    context.Context
+	reader client.Reader
+}
+
+func (p *shootPredicate) InjectStopChannel(stopChan <-chan struct{}) error {
+	p.ctx = contextutil.FromStopChannel(stopChan)
+	return nil
+}
+
+func (p *shootPredicate) InjectClient(client client.Client) error {
+	p.reader = client
+	return nil
+}
+
+func (p *shootPredicate) Create(e event.CreateEvent) bool { return p.filterShoot(e.Object, false) }
+
+func (p *shootPredicate) Update(e event.UpdateEvent) bool {
+	shoot, ok := e.ObjectNew.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return false
+	}
+
+	oldShoot, ok := e.ObjectOld.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return false
+	}
+
+	if !reflect.DeepEqual(shoot.DeletionTimestamp, oldShoot.DeletionTimestamp) || shootHealthStatus(shoot) != shootHealthStatus(oldShoot) {
+		return true
+	}
+
+	return p.filterShoot(e.ObjectNew, false)
+}
+
+func (p *shootPredicate) Delete(e event.DeleteEvent) bool { return p.filterShoot(e.Object, true) }
+
+func (p *shootPredicate) Generic(_ event.GenericEvent) bool { return false }
+
+// Return true only if the Shoot belongs to the pending replica and it progressed from the state
+// that caused the replica to be pending
+func (p *shootPredicate) filterShoot(obj client.Object, deleted bool) bool {
```

Comment:
😵 Can we get rid of this `deleted` value and handle this single case in the `Delete()` handler? Now, we just copied the confusing code but didn't improve it.